### PR TITLE
added sendCommerceEvent

### PR DESCRIPTION
**Pull request by EvangelosG**
added sendCommerceEvent. Usage:

```
   private void sendCommerceEvent() {
        Branch branch = Branch.getInstance();
        Product product = new Product("acme007", "Acme brand 1 ton weight", new BigDecimal(1000.99), 1, "Acme", "Widgets", "Lite Weight" );
        CommerceEvent commerceEvent = new CommerceEvent(new BigDecimal(1101.99), "Smackeroos", "tr00x8", new BigDecimal(100.00), new BigDecimal(1.0),
                "Acme weights coupon", "ACME by Amazon", product);
        JSONObject jsonObject = new JSONObject();
        try { jsonObject.put("metakey", "metavalue"); } catch ( JSONException e ) {}
        branch.sendCommerceEvent(commerceEvent, jsonObject, null);
    }
```

how should I handle decimal places? see CommerceEvent/Product

@agrimn @Sarkar @E-B-Smith

### CODE REVIEW

**Comment by derrickstaten** [comment]
@agrimn @Sarkar can you please review when you have time?

File: Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+        try {
+            post.put(Defines.Jsonkey.IdentityID.getKey(), prefHelper_.getIdentityID());
+            post.put(Defines.Jsonkey.DeviceFingerprintID.getKey(), prefHelper_.getDeviceFingerPrintID());
+            post.put(Defines.Jsonkey.SessionID.getKey(), prefHelper_.getSessionID());
+            if (!prefHelper_.getLinkClickID().equals(PrefHelper.NO_STRING_VALUE)) {
+                post.put(Defines.Jsonkey.LinkClickID.getKey(), prefHelper_.getLinkClickID());
+            }
+            post.put(Defines.Jsonkey.Event.getKey(), "purchase");
+            if (metadata != null)
```
Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java]: Brackets around if statement

File: Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    public ServerRequestRActionCompleted(Context context, CommerceEvent commerceEvent, JSONObject metadata, BranchViewHandler.IBranchViewEvents callback) {
+        super(context, Defines.RequestPath.CompletedAction.getPath());
+        callback_ = callback;
+        JSONObject post = new JSONObject();
+
+        try {
+            post.put(Defines.Jsonkey.IdentityID.getKey(), prefHelper_.getIdentityID());
+            post.put(Defines.Jsonkey.DeviceFingerprintID.getKey(), prefHelper_.getDeviceFingerPrintID());
+            post.put(Defines.Jsonkey.SessionID.getKey(), prefHelper_.getSessionID());
+            if (!prefHelper_.getLinkClickID().equals(PrefHelper.NO_STRING_VALUE)) {
```
Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java]: We want to pass the link click ID in this post? Is this for analytics purposes? cc @derrickstaten -- just confirming this is required

File: Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                String actionName = "";
+                try {
+                    JSONObject post = getPost();
+                    if (post != null && post.has(Defines.Jsonkey.Event.getKey())) {
+                        actionName = post.getString(Defines.Jsonkey.Event.getKey());
+                    }
+                    if (Branch.getInstance().currentActivityReference_ != null) {
+                        Activity currentActivity = Branch.getInstance().currentActivityReference_.get();
+                        JSONObject branchViewJsonObj = resp.getObject().getJSONObject(Defines.Jsonkey.BranchViewData.getKey());
+                        BranchViewHandler.getInstance().showBranchView(branchViewJsonObj, actionName, currentActivity, callback_);
```
Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java]: Why are we showing a branch view here? Is this related to what we want to return when a commerce event is saved?

File: Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java (hunk 2 above)
Comment (derrickstaten) [inline Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java]: I'd say send up whatever is already sent up for all events

File: Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java (hunk 3 above)
Comment (derrickstaten) [inline Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java]: again, we should just imitate what is already present for userCompletedAction

Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java]: ok cool -- if that's the intention this code makes sense to me 👍 🌮

File: Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java (hunk 1 above)
Comment (EvangelosG) [inline Branch-SDK/src/io/branch/referral/ServerRequestRActionCompleted.java]: added!

File: Branch-SDK/src/io/branch/referral/util/CommerceEvent.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        this.currency = currency;
+        this.transactionID = transactionID;
+        this.shipping = shipping;
+        this.tax = tax;
+        this.coupon = coupon;
+        this.affiliation = affiliation;
+        this.products = products;
+    }
+
+    public CommerceEvent(BigDecimal revenue, String currency, String transactionID, BigDecimal shipping, BigDecimal tax, String coupon, String affiliation, Product product) {
```
Comment (sojanpr) [inline Branch-SDK/src/io/branch/referral/util/CommerceEvent.java]: We have currenctType enumeration in SDK which is already used in BUO. The "string" for currency name can be replaced with it

**Comment by dmitrig01** [comment]
right now on the backend we are just handling the revenue as a double. I think we decided to forgo the complexity of more precision for now.

File: Branch-SDK/src/io/branch/referral/util/CommerceEvent.java (hunk 4 above)
Comment (EvangelosG) [inline Branch-SDK/src/io/branch/referral/util/CommerceEvent.java]: thanks for letting me know! I replaced currency w/ currencyType.

**Comment by dmitrig01** [comment]
please see comment about the type -- i think double, not BigDecimal, should be fine for now.

**Comment by EvangelosG** [comment]
```
private void sendCommerceEvent() {
        Branch branch = Branch.getInstance();
        Product product = new Product("acme007", "Acme brand 1 ton weight", 1.0, 1, "Acme", "uh", ProductCategory.ANIMALS_AND_PET_SUPPLIES);
        CommerceEvent commerceEvent = new CommerceEvent(1101.99, CurrencyType.USD, "tr00x8", 100.00, 1.0,
                "Acme weights coupon", "ACME by Amazon", product);
        JSONObject jsonObject = new JSONObject();
        try { jsonObject.put("metakey", "metavalue"); } catch ( JSONException e ) {}
        branch.sendCommerceEvent(commerceEvent, jsonObject, null);
    }
```

**Pull request merged**